### DISCUSSION TITLE
Why is the Average Miss Latency calculated using the fill time of the later merged entry?

### DISCUSSION THREAD

**Discussion by icuuuuuuer** [ChampSim/ChampSim#556] [Q&A] [answered]
I found that Champsim calculates the Average Miss Latency using the fill time of the later merged entry. Could this cause the actual Average Miss Latency to be larger than the result given by Champsim? Here’s an overall summary of this part of the code，I use master version.
```C++
CACHE::mshr_type CACHE::mshr_type::merge(mshr_type predecessor, mshr_type successor)
{
  std::vector<std::reference_wrapper<ooo_model_instr>> merged_instr{};
  std::vector<std::deque<response_type>*> merged_return{};
  std::set_union(std::begin(predecessor.instr_depend_on_me), std::end(predecessor.instr_depend_on_me), std::begin(successor.instr_depend_on_me),
                 std::end(successor.instr_depend_on_me), std::back_inserter(merged_instr), ooo_model_instr::program_order);
  std::set_union(std::begin(predecessor.to_return), std::end(predecessor.to_return), std::begin(successor.to_return), std::end(successor.to_return),
                 std::back_inserter(merged_return));
  mshr_type retval{(successor.type == access_type::PREFETCH) ? predecessor : successor};
  retval.instr_depend_on_me = merged_instr;
  retval.to_return = merged_return;
  retval.data = predecessor.data;
  return retval;
}
```
This part of the code shows that if there are two load requests for the same cacheline, load1 arriving before load2, load2 will merge load1 in the MSHR.

The Average Miss Latency is calculated as the total cache miss latency divided by the total number of misses. The total miss latency is determined by subtracting the MSHR enqueue cycle from the fill cycle. However, if load2 merges load1, the observed enqueue cycle during the fill will be that of load2, not load1.

This can result in a smaller total miss latency, which in turn makes the Average Miss Latency smaller.
```C++
bool CACHE::handle_fill(const mshr_type& fill_mshr)
{
//....
  if (success) {
    // COLLECT STATS
    sim_stats.total_miss_latency += current_cycle - (fill_mshr.cycle_enqueued + 1);
}}



bool CACHE::handle_miss(const tag_lookup_type& handle_pkt)
{
 mshr_type to_allocate{handle_pkt, current_cycle};
//...
  // check mshr
  auto mshr_entry = std::find_if(std::begin(MSHR), std::end(MSHR), [match = handle_pkt.address >> OFFSET_BITS, shamt = OFFSET_BITS](const auto& entry) {
    return (entry.address >> shamt) == match;
  });
  bool mshr_full = (MSHR.size() == MSHR_SIZE);
  if (mshr_entry != MSHR.end()) // miss already inflight
  {
    if (mshr_entry->type == access_type::PREFETCH && handle_pkt.type != access_type::PREFETCH) {
      // Mark the prefetch as useful
      if (mshr_entry->prefetch_from_this)
        ++sim_stats.pf_useful; 
    }
    *mshr_entry = mshr_type::merge(*mshr_entry, to_allocate);
  } }

```
Could you kindly explain why the later entry is used to merge the original entry? What is the purpose of this design?
Additionally, I find the prefetch_from_this signal quite confusing. Could you please provide a brief explanation of what this signal represents?Thank you so much for your help!

**Comment by maccoymerrell** [comment, marked as answer]
See PR #536 regarding miss latency and handling MSHR merges.
The key reason that is being done is to preserve the original type of the MSHR when prefetches are merged into demands.
It has the negative consequence of altering the 
measured latency, which is again addressed by the pr.

Regarding prefetch_from_this bool, I believe that just indicates whether the given prefetch is from that particular cache's prefetcher (true), or is instead from an upstream prefetcher (false). For stat collecting purposes, prefetcher usefulness metrics try to only collect data for that particular cache's prefetcher.

**Reply by icuuuuuuer** [reply]
Thank you! I reviewed the average miss latency in the develop version, and it's absolutely correct.
